httpclient/metrics_round_tripper.go

```diff
@@ -0,0 +1,121 @@
+		}, []string{"client_type", "remote_address", "summary", "status"}),
+	}
+}
+
+// MustRegister registers the Prometheus metrics.
+func (p *PrometheusMetricsCollector) MustRegister() {
+	prometheus.MustRegister(p.Durations)
+}
+
+// RequestDuration observes the duration of the request and the status code.
+func (p *PrometheusMetricsCollector) RequestDuration(requestType, host, summary, status string, start time.Time) {
+	p.Durations.WithLabelValues(requestType, host, summary, status).Observe(time.Since(start).Seconds())
+}
+
+// Unregister the Prometheus metrics.
+func (p *PrometheusMetricsCollector) Unregister() {
+	prometheus.Unregister(p.Durations)
+}
+
+// MetricsRoundTripper is an HTTP transport that measures requests done.
+type MetricsRoundTripper struct {
+	// Delegate is the next RoundTripper in the chain.
+	Delegate http.RoundTripper
+
+	// ClientType represents a type of client, it's a service component reference. e.g. 'auth-service'
+	ClientType string
+
+	// Collector is a metrics collector.
+	Collector MetricsCollector
+
+	// ClassifyRequest does request classification, producing non-parameterized summary for given request.
+	ClassifyRequest func(r *http.Request, requestType string) string
+}
+
+// MetricsRoundTripperOpts is an HTTP transport that measures requests done.
+type MetricsRoundTripperOpts struct {
+	// ClientType represents a type of client, it's a service component reference. e.g. 'auth-service'
+	ClientType string
+
+	// ClassifyRequest does request classification, producing non-parameterized summary for given request.
+	ClassifyRequest func(r *http.Request, requestType string) string
+}
+
+// NewMetricsRoundTripper creates an HTTP transport that measures requests done.
+func NewMetricsRoundTripper(delegate http.RoundTripper, collector MetricsCollector) http.RoundTripper {
+	return NewMetricsRoundTripperWithOpts(delegate, collector, MetricsRoundTripperOpts{})
+}
+
+// NewMetricsRoundTripperWithOpts creates an HTTP transport that measures requests done.
+func NewMetricsRoundTripperWithOpts(
+	delegate http.RoundTripper,
+	collector MetricsCollector,
+	opts MetricsRoundTripperOpts,
+) http.RoundTripper {
+	return &MetricsRoundTripper{
+		Delegate:        delegate,
+		ClientType:      opts.ClientType,
+		Collector:       collector,
+		ClassifyRequest: opts.ClassifyRequest,
+	}
+}
+
+// RoundTrip measures external requests done.
+func (rt *MetricsRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
+	if rt.Collector == nil {
+		return rt.Delegate.RoundTrip(r)
+	}
+
+	status := "0"
+	start := time.Now()
+
+	resp, err := rt.Delegate.RoundTrip(r)
+	if err == nil && resp != nil {
+		status = strconv.Itoa(resp.StatusCode)
+	}
+
+	summary := fmt.Sprintf("%s %s", r.Method, rt.ClientType)
+	if rt.ClassifyRequest != nil {
+		summary = rt.ClassifyRequest(r, rt.ClientType)
+	}
+
+	rt.Collector.RequestDuration(rt.ClientType, r.Host, summary, status, start)
```

Comment:
In this implementation, the request’s duration will include the time taken by `rt.ClassifyRequest`, which means it depends on how the developer implements it. How about measuring only the time taken by `rt.Delegate.RoundTrip(r)` and passing this calculated duration to `rt.Collector.RequestDuration()`? This approach would provide a more accurate measurement of the actual request duration.